### PR TITLE
Better handling of empty (not missing) docstrings

### PR DESCRIPTION
**Pull request by dzbarsky**
#### What does this implement/fix?
Some python build tools/processes set docstrings to empty strings. (Not -OO). This can be beneficial to work around various libraries depending on docstrings existing but not caring about their value. Some places in scipy already work with empty string, this aligns some existing handling that wasn't consistent.

### CODE REVIEW

**Comment by ev-br** [comment]
Can you add some more info on what is the motivation for these changes? Does somebody/something really reach into the library and fiddle with `__doc__` attributes?

**Comment by dzbarsky** [comment]
> Can you add some more info on what is the motivation for these changes? Does somebody/something really reach into the library and fiddle with `__doc__` attributes?

At Dropbox we deploy a monolithic application. Setting docstrings to empty strings saves 5%+ from memory usage, but we do it selectively for modules (it's enabled for 99% of modules) Using -OO is not an option because we cannot enable it for all modules, and stripping assertions will break the application. Currently we are not stripping docstrings for scipy but would like to remove this exclusion from our list.

**Comment by ev-br** [comment]
Thanks.
Could you rebase on master, to see if the CI failures go away.

File: scipy/integrate/quadrature.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ def _copy_func(f):
 
 
 trapz = _copy_func(trapz)
-if sys.flags.optimize <= 1:
+if sys.flags.optimize <= 1 and trapz.__doc__:
```
Comment (peterbell10) [inline scipy/integrate/quadrature.py]:
```suggestion
if trapz.__doc__:
```

**Review by peterbell10 — COMMENTED** [review]
> At Dropbox we deploy a monolithic application. Setting docstrings to empty strings saves 5%+ from memory usage

Almost everything should work with `__doc__ = None` already. So is the main issue that you're using empty strings instead of `None`?

**Comment by dzbarsky** [comment]
> > At Dropbox we deploy a monolithic application. Setting docstrings to empty strings saves 5%+ from memory usage
> 
> Almost everything should work with `__doc__ = None` already. So is the main issue that you're using empty strings instead of `None`?

When editing the parsed AST, docstrings must be set to a string, not None. So yes, we set them to empty strings. (Another approach is to completely remove docstring nodes, but this is more complex cause then you need to insert `pass` statements).

**Comment by rgommers** [comment]
The CI failures look real

File: scipy/optimize/nonlin.py [hunk 2]
```diff
@@ -217,8 +218,7 @@ def _safe_norm(v):
 
 
 def _set_doc(obj):
-    if obj.__doc__:
-        obj.__doc__ = obj.__doc__ % _doc_parts
```
Comment (larsoner) [inline scipy/optimize/nonlin.py]: This seems like it will break when `obj.__doc__ is None`, is this change necessary?

**Comment by larsoner** [comment]
FYI @dzbarsky in case you want to get this into 1.4.0, we are going to branch in under two weeks

File: scipy/optimize/nonlin.py (hunk 2 above)
Comment (larsoner) [reply scipy/optimize/nonlin.py]: (Also this is likely the source of the CircleCI failure)

**Comment by dzbarsky** [comment]
@larsoner updated. The only failing test is https://github.com/scipy/scipy/blob/c437dc0e84d8b7c4942cdff6a1eb99430e47e134/scipy/linalg/tests/test_decomp.py#L2472-L2484 which I don't see how it's affected. Is it a flake? Could we retry/merge this PR if it passes? Hopefully it makes the 1.4 branch!

**Comment by rgommers** [comment]
All green, in it goes. Thanks @dzbarsky.
